File: app_java/src/main/java/vlab/android/architecture/feature/login/viewmodel/LoginViewModel.java
package vlab.android.architecture.feature.login.viewmodel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;

import javax.inject.Inject;

import vlab.android.architecture.base.BaseViewModel;
import vlab.android.architecture.feature.login.model.UserInfo;
import vlab.android.architecture.feature.login.usecase.LoginUseCase;

/**
 * Created by Vinh Tran on 2/15/18.
 */

public class LoginViewModel extends BaseViewModel {

    private LoginUseCase mLoginUseCase;

    @Inject
    public LoginViewModel(LoginUseCase loginUseCase) {
        mLoginUseCase = loginUseCase;
    }

    @Override
    protected void onCleared() {
        mLoginUseCase.onCleared();
        super.onCleared();
    }

    public void login(String userName, String pwd) {
        // map request data
        LoginUseCase.LoginRequestParam requestData = new LoginUseCase.LoginRequestParam(userName, pwd);

        mLoginUseCase.login(requestData);
    }

    public void cancelLogin(){
        mLoginUseCase.cancelLoginRequest();
    }

    public LiveData<UserInfo> onLoginSuccessObs() {
        return Transformations.map(mLoginUseCase.onLoginSuccessObs(), UserInfo::new);
    }

    public LiveData<Throwable> onLoginErrorObs() {
        return mLoginUseCase.onLoginErrorObs();
    }

    public LiveData<Boolean> onLoadingObs() {
        return mLoginUseCase.onLoadingObs();
    }
}

File: app_kotlin/src/main/java/vlab/android/architecture/feature/login/viewmodel/LoginViewModel.kt
package vlab.android.architecture.feature.login.viewmodel

import android.arch.lifecycle.LiveData
import vlab.android.architecture.base.BaseViewModel
import vlab.android.architecture.feature.login.use_case.LoginUseCase
import vlab.android.architecture.feature.user.UserInfo
import javax.inject.Inject

/**
 * Created by Vinh Tran on 2/15/18.
 */

class LoginViewModel @Inject
constructor(private val mLoginUseCase: LoginUseCase) : BaseViewModel() {

    override fun onCleared() {
        mLoginUseCase.onCleared()
        super.onCleared()
    }

    fun login(userName: String, pwd: String) {
        // map request data
        val requestData = LoginUseCase.LoginRequestParam(userName, pwd)

        mLoginUseCase.login(requestData)
    }

    fun cancelLogin() {
        mLoginUseCase.cancelLoginRequest()
    }

    fun onLoginSuccessObs(): LiveData<UserInfo> {
        return mLoginUseCase.onLoginSuccessObs()
    }

    fun onLoginErrorObs(): LiveData<Throwable> {
        return mLoginUseCase.onLoginErrorObs()
    }

    fun onLoadingObs(): LiveData<Boolean> {
        return mLoginUseCase.onLoadingObs()
    }
}


